src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs

```diff
@@ -148,15 +148,21 @@ private static HttpConnectionKey GetConnectionKey(HttpRequestMessage request, Ur
         {
             Uri uri = request.RequestUri;
 
+            // TODO: #28863 Part 1: Uri.IdnHost should include [] around IPv6 address.
             // If the hostname is an IPv6 address, uri.IdnHost will return the address without enclosing [].
             // In this case, use uri.Host instead, which will correctly enclose with [].
             // Note we don't need punycode encoding if it's an IP address, so using uri.Host is fine.
-            bool isIPv6Address = uri.HostNameType == UriHostNameType.IPv6;
+
+            // TODO: #28863 Part 2: Uri.Host LLA (Link-local address) for IPv6 address doesn't contain %number part.
+            // Since scope is mandatory for LLA, we will use uri.IdnHost for IPv6 LLA address.
+            // From RFC 4291, IPv6 LLA has formate like: fe80::/10.
+            bool isNonLLAIPv6Address = (uri.HostNameType == UriHostNameType.IPv6 &&
+                uri.IdnHost.IndexOf("fe80") != 0);
 
             if (isProxyConnect)
             {
                 Debug.Assert(uri == proxyUri);
-                return new HttpConnectionKey(HttpConnectionKind.ProxyConnect, isIPv6Address ? uri.Host : uri.IdnHost, uri.Port, null, proxyUri);
+                return new HttpConnectionKey(HttpConnectionKind.ProxyConnect, isNonLLAIPv6Address ? uri.Host : uri.IdnHost, uri.Port, null, proxyUri);
```

Comment:
Can you remind me why we need to distinguish between Host and IdnHost in these connection keys?  Is the host from the connection key ever written out anywhere?  It looks like "no", in which case why does it matter whether it's wrapped in []?

Reply:
> Can you remind me why we need to distinguish between Host and IdnHost in these connection keys? Is the host from the connection key ever written out anywhere?

The host from the connection key will be written out to [the `Host` header](https://github.com/caesar1995/corefx/blob/22633ac67e1b59ddc058b7e1d700e90e5c202934/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L267).

We need to distinguish because `IdnHost` doesn't include `[]` around IPv6 address, thus the `Host` header we are sending out for IPv6 address will be missing `[]`. (#28557)

> in which case why does it matter whether it's wrapped in []?

As the example in #28557 showed, if it not wrapped in `[]`, for example, `Host: ::1:5001`. For IPv6 address, it will be hard to tell which part is address, which part is port number.

Reply:
> The host from the connection key will be written out to the Host header.

It doesn't have to be though. This code is executed for every request. In contrast, this code is executed only when we create a new pool:
https://github.com/caesar1995/corefx/blob/22633ac67e1b59ddc058b7e1d700e90e5c202934/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPoolManager.cs#L211
and it's that ctor that creates the bytes to be written out. We can pass whatever we want into that ctor, such as the host bytes to use or something the ctor can use to compute the host bytes. We should be keeping the connection key logic as lean as possible.

Reply:
> We should be keeping the connection key logic as lean as possible.

I agree. But the tricky part here is to determine the value we want to pass into that ctor without additional allocation. I will explore this approach to see if able to find efficient way to do so, so that we can make a trade-off decision.

Reply:
An extra string allocation when constructing the pool is fine if there's no better way.  It's not fine when constructing a key.